zerver/lib/markdown/__init__.py

```diff
@@ -615,18 +616,21 @@ def add_a(
         if data_id is not None:
             a.set("data-id", data_id)
         img = SubElement(a, "img")
-        if (
-            settings.THUMBNAIL_IMAGES
-            and (not already_thumbnailed)
-            and user_uploads_or_external(image_url)
-        ):
-            # We strip leading '/' from relative URLs here to ensure
-            # consistency in what gets passed to /thumbnail
-            image_url = image_url.lstrip("/")
-            img.set("src", "/thumbnail?" + urlencode({"url": image_url, "size": "thumbnail"}))
-            img.set(
-                "data-src-fullsize", "/thumbnail?" + urlencode({"url": image_url, "size": "full"})
-            )
+        if image_url.startswith("/user_uploads/") and self.zmd.zulip_db_data:
+            path_id = image_url[len("/user_uploads/") :]
+            if path_id in self.zmd.zulip_db_data.user_upload_previews:
+                # Insert a placeholder image spinner.  We post-process
+                # this content (see rewrite_thumbnailed_images in
+                # zerver.lib.thumbnail), looking specifically for this
+                # tag, and may re-write it into the thumbnail URL if it
+                # already exists when the message is sent.
+                img.set("class", "image-loading-placeholder")
+                img.set("src", "/static/images/loading/loader-black.svg")
+            else:
+                # This shouldn't be possible, based on our check in
+                # is_image for the path_id being in
+                # user_upload_previews; for safety, treat as non-thumbnailed
+                img.set("src", image_url)
```

Comment:
Would an assertion error be more appropriate here?

Reply:
Mmmmmmaybe?  I guess I'll swap one in.